File: week3Story/Story_3/Task4.java
/*
 * Filename  - Task4.java
 * Description - You are given two singly linked lists, head1 and head2, and an integer k. Write a
        function kZiplineMerge(head1, head2, k) that merges the two lists by "zipping" them in
        chunks of k. You should take k nodes from head1, then k nodes from head2, then the next k
        nodes from head1, and so on.
        If one list runs out of nodes, append the remainder of the other list's current chunk, and then
        append the rest of the other list. This must be done in-place by modifying pointers.
 * Author - Nancy Jain
 */import java.util.*;

public class Task4 {

    public static int readInt(Scanner sc) {
        while (true) {
            try {
                return Integer.parseInt(sc.next());
            } catch (Exception e) {
                System.out.print("Invalid input! Enter again: ");
            }
        }
    }

    static class Node {
        char value;
        Node next;

        Node(char value) {
            this.value = value;
            this.next = null;
        }
    }

    static Node buildList(char[] arr) {
        Node head = null, tail = null;
        for (char ch : arr) {
            Node newNode = new Node(ch);
            if (head == null) {
                head = tail = newNode;
            } else {
                tail.next = newNode;
                tail = newNode;
            }
        }
        return head;
    }

    static void printList(Node head) {
        while (head != null) {
            System.out.print(head.value);
            if (head.next != null) System.out.print(" -> ");
            head = head.next;
        }
        System.out.println();
    }

    static Node extractKNodes(Node[] headRef, int k) {
        Node head = headRef[0];
        if (head == null) return null;
        Node start = head;
        Node last = null;
        while (head != null && k-- > 0) {
            last = head;
            head = head.next;
        }
        if (last != null) last.next = null;
        headRef[0] = head;   
        return start;
    }

    static Node kZipMerge(Node head1, Node head2, int k) {
        Node mergedHead = null, mergedTail = null;
        boolean takeFirst = true;
        Node[] ref1 = {head1};
        Node[] ref2 = {head2};
        while (ref1[0] != null || ref2[0] != null) {
            Node chunk = takeFirst ? extractKNodes(ref1, k): extractKNodes(ref2, k);
            takeFirst = !takeFirst;
            if (chunk == null) continue;
            if (mergedHead == null) {
                mergedHead = mergedTail = chunk;
            } else {
                mergedTail.next = chunk;
            }
            while (mergedTail.next != null) {
                mergedTail = mergedTail.next;
            }
        }
        return mergedHead;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (true) {
            System.out.print("Enter size of List 1: ");
            int n1 = readInt(sc);
            char[] arr1 = new char[n1];
            System.out.println("Enter elements of List 1: ");
            for (int i = 0; i < n1; i++) {
                arr1[i] = sc.next().charAt(0);
            }
            System.out.print("Enter size of List 2: ");
            int n2 = readInt(sc);
            char[] arr2 = new char[n2];
            System.out.println("Enter elements of List 2: ");
            for (int i = 0; i < n2; i++) {
                arr2[i] = sc.next().charAt(0);
            }
            System.out.print("Enter k: ");
            int k = readInt(sc);
            Node head1 = buildList(arr1);
            Node head2 = buildList(arr2);
            System.out.println("\nList 1:");
            printList(head1);
            System.out.println("List 2:");
            printList(head2);
            Node merged = kZipMerge(head1, head2, k);
            System.out.println("\nMerged (K-Zipline) List:");
            printList(merged);

            System.out.print("\nDo you want to continue? (yes-1 / no-0): ");
            int ch = readInt(sc);
            if (ch == 0) {
                System.out.println("Exiting...");
                break;
            }
        }
    }
}

File: Story 3/Task4.cpp
/**
 * @filename - Task-4.cpp
 * @description - The K-Zipline Merge 
 * @author - Disha Toshniwal 
 */
#include <iostream>
#include <vector>
using namespace std;

struct Node {
    int val;
    Node* next;
    Node(int v) : val(v), next(nullptr) {}
};

Node* buildList(const vector<int>& arr) {
    Node* head = nullptr;
    Node* tail = nullptr;
    for (int x : arr) {
        Node* n = new Node(x);
        if (!head) head = tail = n;
        else tail = tail->next = n;
    }
    return head;
}

void printList(Node* head) {
    while (head) {
        cout << head->val << (head->next ? " " : "");
        head = head->next;
    }
    cout << "\n";
}

Node* takeK(Node*& head, int k) {
    if (!head) return nullptr;
    Node* start = head;
    Node* prev = nullptr;
    while (head && k--) {
        prev = head;
        head = head->next;
    }
    if (prev) prev->next = nullptr; 
    return start;
}

Node* kZiplineMerge(Node* a, Node* b, int k) {
    Node* result = nullptr;
    Node* tail = nullptr;
    bool turnA = true;

    while (a || b) {
        Node* chunk = turnA ? takeK(a, k) : takeK(b, k);
        turnA = !turnA;

        if (!chunk) continue; 

        if (!result) result = tail = chunk;
        else {
            tail->next = chunk;
        }

        while (tail->next) tail = tail->next; 
    }

    return result;
}

int main() {
    int n1, n2, k;

    cout << "Enter size of list1: ";
    cin >> n1;
    vector<int> v1(n1);
    cout << "Enter list1 values:\n";
    for (int& x : v1) cin >> x;

    cout << "Enter size of list2: ";
    cin >> n2;
    vector<int> v2(n2);
    cout << "Enter list2 values:\n";
    for (int& x : v2) cin >> x;

    cout << "Enter k: ";
    cin >> k;

    Node* head1 = buildList(v1);
    Node* head2 = buildList(v2);

    cout << "\nList1: ";
    printList(head1);
    cout << "List2: ";
    printList(head2);

    Node* merged = kZiplineMerge(head1, head2, k);

    cout << "\nMerged List: ";
    printList(merged);

    return 0;
}
